internal/handlers/common/findandmodify.go

```diff
@@ -168,3 +178,152 @@ func GetFindAndModifyParams(doc *types.Document, l *zap.Logger) (*FindAndModifyP
 		MaxTimeMS:          maxTimeMS,
 	}, nil
 }
+
+// UpsertDocument updates the first document if exists, or create an insert document
+// from the params if no documents in query result docs.
+func UpsertDocument(docs []*types.Document, params *FindAndModifyParams) (*UpsertParams, error) {
```

Comment:
Same here: this function doesn't upsert anything.

Reply:
Renamed to `PrepareDocumentForUpsert`